### PR TITLE
feat: Handle new `avvio` error codes

### LINKED ISSUE #2842: Improve errors user experience
Fastify has many assertions around in the codebase to make sure that the user didn't misconfigure something, for instance, it checks if a route has been declared twice or if a plugin is not a function.
This is great, we vastly reduce the possibility of weird runtime issues, but for historical reasons the errors we are emitting are everything but user friendly.

We should improve this situation, as it is a big pain point in our user experience.

Example of double route declaration:
```js
'use strict'
const fastify = require('fastify')()
fastify.get('/', () => {})
fastify.get('/', () => {})
fastify.listen(3000, console.log)
```

```
▶ node index.js
AssertionError [ERR_ASSERTION]: Method 'GET' already declared for route '/'
    at Router._insert (.../fst-err/node_modules/find-my-way/index.js:304:9)
    at Router._on (.../fst-err/node_modules/find-my-way/index.js:197:8)
    at Router.on (.../fst-err/node_modules/find-my-way/index.js:74:8)
    at Object.afterRouteAdded (.../fst-err/node_modules/fastify/lib/route.js:253:16)
    at .../fst-err/node_modules/fastify/lib/route.js:187:25
    at Object._encapsulateThreeParam (.../fst-err/node_modules/avvio/boot.js:545:7)
    at Boot.timeoutCall (.../fst-err/node_modules/avvio/boot.js:447:5)
    at Boot.callWithCbOrNextTick (.../fst-err/node_modules/avvio/boot.js:428:19)
    at Boot._after (.../fst-err/node_modules/avvio/boot.js:273:26)
    at Plugin.exec (.../fst-err/node_modules/avvio/plugin.js:131:17) {
  generatedMessage: false,
  code: 'ERR_ASSERTION',
  actual: false,
  expected: true,
  operator: '=='
}
```

What I would expect instead is:
```
FastifyError: Method 'GET' already declared for route '/'
   at fastify.get('/', () => {}) (.../fst-err/index.js:4:9)
```
Which is quite complex, given that our route declaration is asynchronous.

Another example, this time with bad plugin type:
```js
'use strict'
const fastify = require('fastify')()
fastify.register(null)
fastify.listen(3000, console.log)
```
```
▶ node index.js
.../fst-err/node_modules/avvio/boot.js:200
    throw new Error('plugin must be a function or a promise')
    ^

Error: plugin must be a function or a promise
    at assertPlugin (.../fst-err/node_modules/avvio/boot.js:200:11)
    at Boot._addPlugin (.../fst-err/node_modules/avvio/boot.js:233:12)
    at Boot.use (.../fst-err/node_modules/avvio/boot.js:209:25)
    at Object.server.<computed> [as register] (.../fst-err/node_modules/avvio/boot.js:33:14)
    at Object.<anonymous> (.../fst-err/index.js:5:9)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
```

As you can see, the error is not being helpful at all.

### PR DESCRIPTION
**Pull request by MetCoder95**
Closes #2842

**Note**:
- This PR depends on the next major release of `avvio` until then is pointing out to `fastify/avvio#next` branch for dependency.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Review by jsumners — APPROVED** [review]
Looks good to me. Just need to rebase a bit.

File: fastify.js [hunk 1]
```diff
@@ -462,6 +463,10 @@ function fastify (options) {
     }
 
     function manageErr (err) {
+      err = err != null && AVVIO_ERRORS_MAP[err.code] != null
+        ? new AVVIO_ERRORS_MAP[err.code](err.message)
```
[the remark this answers is not in the document]
Comment (jsumners) [reply fastify.js:467]:
```js
const newError = Error('foo')
newError.cause = originalError
```

File: test/plugin.test.js [hunk 2]
```diff
@@ -845,8 +845,8 @@ test('plugin metadata - dependencies (nested)', t => {
   }
 })
 
-test('pluginTimeout', t => {
-  t.plan(2)
+test('pluginTimeout', { only: true }, t => {
```
Comment (Eomm) [inline test/plugin.test.js:848]:
```suggestion
test('pluginTimeout', t => {
```

File: test/plugin.test.js [hunk 3]
```diff
@@ -855,12 +855,14 @@ test('pluginTimeout', t => {
   })
   fastify.ready((err) => {
     t.ok(err)
-    t.equal(err.code, 'ERR_AVVIO_PLUGIN_TIMEOUT')
+    t.equal(err.message,
+      "fastify-plugin: Plugin did not start in time: 'function (app, opts, done) { -- // to no call done on purpose'. You may have forgotten to call 'done' function or to resolve a Promise")
```
Comment (Eomm) [inline test/plugin.test.js:859]: Could we add a test with a named function too?

File: fastify.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (MetCoder95) [reply fastify.js]: I followed @jsumners approach, applied in 27ce870

File: test/hooks.on-ready.test.js [hunk 4]
```diff
@@ -277,7 +277,7 @@ t.test('onReady cannot add lifecycle hooks', t => {
       fastify.addHook('onRequest', (request, reply, done) => {})
     } catch (error) {
       t.ok(error)
-      t.equal(error.message, 'root plugin has already booted')
+      t.equal(error.message, 'Root plugin has already booted')
```
[the remark this answers is not in the document]
Comment (MetCoder95) [reply test/hooks.on-ready.test.js:280]: Sure! done in cbb0e73

File: test/plugin.test.js (hunk 3 above)
Comment (MetCoder95) [reply test/plugin.test.js:859]: Sure, made a new one in cbb0e73

File: docs/Errors.md [hunk 5]
```diff
@@ -173,3 +173,33 @@ You cannot use `send` inside the `onError` hook.
...
+
+<a name="FST_ERR_PLUGIN_NOT_VALID"></a>
+#### FST_ERR_PLUGIN_NOT_VALID
+
+Plugin must be a function or a promise.
+
+<a name="FST_ERR_PLUGIN_TIMEOUT"></a>
+#### FST_ERR_PLUGIN_TIMEOUT
+
+Plugin did not start in time.
```
Comment (RafaelGSS) [inline docs/Errors.md:185]: Could we share the timeout default here?

File: test/hooks.on-ready.test.js [hunk 6]
```diff
@@ -277,7 +277,9 @@ t.test('onReady cannot add lifecycle hooks', t => {
...
+      // TODO: look where the error pops up
```
[the remark this answers is not in the document]
Comment (MetCoder95) [reply test/hooks.on-ready.test.js:281]: Yes, I'm taking a look right now. I just wanted to push the changes already made for keeping track of the progress 👍 

Also, this is quite weird. The thing is that it seems it throws sync, and for instance, I'm not able to fully catch it, do you have any idea what I'm missing?

File: docs/Errors.md (hunk 5 above)
Comment (MetCoder95) [reply docs/Errors.md]: Done in d315fed

**Pull request merged**